File: medium/tree/PathSum.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode() {}
 *     TreeNode(int val) { this.val = val; }
 *     TreeNode(int val, TreeNode left, TreeNode right) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */
/**
 * 使用了一个 sumList 维护当前层次每次累加数组
 */
class Solution {
  int count;
  public int pathSum(TreeNode root, int sum) {
    count = 0;

    ArrayList<Integer> sumList = new ArrayList<Integer>();
    dfs(root, sum, sumList);
    return count;
  }

  void dfs (TreeNode root, int sum, ArrayList<Integer> sumList) {
    if (root == null) {
      return;
    }
    for (int i = 0; i < sumList.size(); i++) {
      int curSum = sumList.get(i) + root.val;
      sumList.set(i, curSum);
      if (curSum == sum) {
        count++;
      }
    }
    sumList.add(root.val);
    if (root.val) {
      count++;
    }
    dfs(root.left, sum, (ArrayList<Integer>) sumList.clone());
    dfs(root.right, sum, (ArrayList<Integer>) sumList.clone());
  }
}

/**
 * 傻傻的双重dfs
 */
class SolutionDoubleDfs {
  int count;
  public int pathSum(TreeNode root, int sum) {
    count = 0;
    dfs(root, sum);
    return count;
  }

  void dfs (TreeNode root, int sum) {
    if (root == null) {
      return;
    }
    dfsCalculate(root, sum, 0);

    dfs(root.left, sum);
    dfs(root.right, sum);
  }

  void dfsCalculate (TreeNode root, int sum, int accumulate) {
    if (root == null) {
      return;
    }
    accumulate += root.val;
    if (accumulate == sum) {
      count++;
    }
    dfsCalculate(root.left, sum, accumulate);
    dfsCalculate(root.right, sum, accumulate);
  }
}

File: medium/tree/pathSum.ts
// 113. 路径总和 II
// https://leetcode-cn.com/problems/path-sum-ii/
/**
 * Definition for a binary tree node.
 * class TreeNode {
 *     val: number
 *     left: TreeNode | null
 *     right: TreeNode | null
 *     constructor(val?: number, left?: TreeNode | null, right?: TreeNode | null) {
 *         this.val = (val===undefined ? 0 : val)
 *         this.left = (left===undefined ? null : left)
 *         this.right = (right===undefined ? null : right)
 *     }
 * }
 */

function pathSum(root: TreeNode | null, targetSum: number): number[][] {
  const res: number[][] = []

  const route: number[] = []
  function dfs (root: TreeNode | null, currentSum: number) {
    if (!root) {
      return
    }
    const nextSum = currentSum + root.val
    route.push(root.val)
    // 到达叶子节点
    if (!root.left && !root.right) {
      if (nextSum === targetSum) {
        res.push([...route])
      }
      return
    }
    dfs(root.left, nextSum)
    if (root.left) {
      route.pop()
    }
    dfs(root.right, nextSum)
    if (root.right) {
      route.pop()
    }
  }

  dfs(root, 0)

  return res
};

